torch/distributed/fsdp/utils.py

```diff
@@ -8,18 +9,21 @@
 
 
 def _apply_to_tensors(
-    fn: Callable, container: Union[torch.Tensor, Dict, List, Tuple, Set]
+    fn: Callable, container: Union[torch.Tensor, Dict, List, Tuple, Set, OrderedDict, PackedSequence]
 ) -> Any:
     """Recursively apply to all tensor in different kinds of container types."""
 
-    def apply(x: Union[torch.Tensor, Dict, List, Tuple, Set, OrderedDict]) -> Any:
+    def apply(x: Union[torch.Tensor, Dict, List, Tuple, Set, OrderedDict, PackedSequence]) -> Any:
```

Comment:
curious, do we need to support custom types (e.g., by allowing users to pass in a function or by pickling `x` if it is none of the listed types here?)?

Reply:
the apply_to_tensors is mainly used for casting input tensors to mixed precision or register hooks of output tensors internally inside FSDP. possibly it is hard to expose to users for passing a function or so. We also do not want to pickling X for non-tensor data, e..g, we do not want to cast scalars or register hooks of scalars. 

but I'm curious how PyTorch generally get tensors from a custom defined type...? if users' input or output is a custom defined type and it contains tensors, we may not "apply_to_tensors" as expected for this custom defined type right now

Reply:
> but I'm curious how PyTorch generally get tensors from a custom defined type...? if users' input or output is a custom defined type and it contains tensors, we may not "apply_to_tensors" as expected for this custom defined type right now

FWIW, there is no good solution today. What CUDA Future does is pickling the Python obj and extract all tensors during the pickling process, but that can be pretty slow.